code/game/data_huds.dm

```diff
@@ -579,3 +556,28 @@ Diagnostic HUDs!
 	holder.loc = get_turf(src)
 	SET_PLANE(holder,ABOVE_LIGHTING_PLANE,src)
 	set_hud_image_active(MALF_APC_HUD)
+
+/atom/proc/get_cached_width()
+	SHOULD_NOT_OVERRIDE(TRUE)
+	if (isnull(icon))
+		return null
+	if (!GLOB.icon_size_cache[icon])
+		cache_icon_size()
+	return GLOB.icon_size_cache[icon][1]
+
+/atom/proc/get_cached_height()
+	SHOULD_NOT_OVERRIDE(TRUE)
+	if (isnull(icon))
+		return null
+	if (!GLOB.icon_size_cache[icon])
+		cache_icon_size()
+	return GLOB.icon_size_cache[icon][2]
+
+/atom/proc/cache_icon_size()
+	var/icon/sample_icon
+	if (isfile(icon))
+		sample_icon = icon(icon, icon_state, dir)
+	else
+		sample_icon = icon
+	// If icon is a file, we are guaranteed to always have a consistent sprite size for all objects from that icon
+	GLOB.icon_size_cache[isfile(icon) ? icon : sample_icon] = list(sample_icon.Width(), sample_icon.Height())
```

Comment:
worth noting icon_states calls are at least partially internally cached, it's not 100% that this is worthwhile. likely is but wanted to note

Reply:
having some issues with tracy, will post profiler results slightly later